enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,461 @@
+
+# Generic Device Plugin for MicroShift
+
+## Summary
+
+[Generic Device Plugin](https://github.com/squat/generic-device-plugin) is a
+solution for attaching devices not needing special initialization
+(e.g. serial, USB, certain video devices) to containers. The goal of this enhancement is to
+outline its adoption for MicroShift.
+
+## Motivation
+
+MicroShift is built for edge devices, which commonly use peripherals
+like sensors and adapters.
+Currently, MicroShift does not provide any solution to pass these peripherals
+through to Pods without requiring elevated privileges.
+We want to provide a solution for attaching simple devices (not requiring any special
+initialization) to Pods without potentially compromising on security.
+
+### User Stories
+
+* As a MicroShift device administrator, I want to attach simple devices to Pods
+  without elevating privileges (e.g. mounting `/dev`, adding extraneous capabilities).
+
+### Goals
+
+* Include functionality of
+  [Generic Device Plugin](https://github.com/squat/generic-device-plugin)
+  in MicroShift
+
+### Non-Goals
+
+* Adapt Generic Device Plugin (e.g. create Operator) for OpenShift or other Kubernetes distribution
+
+## Proposal
+
+* Import Generic Device Plugin code into MicroShift binary
+* Integrate the Generic Device Plugin's configuration into MicroShift config
+  and add a toggle option
+
+### Workflow Description
+
+**User** is a human user responsible for setting up and managing Edge Devices.
+
+1. The user enables the Generic Device Plugin using MicroShift configuration
+1. Optionally, the user modifies the configuration of Generic Device Plugin, 
+   for example changes device groups and paths
+1. The user (re)starts MicroShift
+1. MicroShift starts the Generic Device Plugin controller which creates plugin
+   goroutines for each device group.
+1. Plugin goroutines detect the devices and report them with the Kubelet
+1. Kubelet updates the Node's status with new devices for scheduler to be aware of Node's capabilities
+1. User deploys a workload utilizing some device
+   (could be at any point, it won't be scheduled/started until the device is available)
+
+In order for workload to get the device attached, the **container** specification
+must include the device in its limits. Below is an example of StatefulSet which
+references a device in a container's limits.
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: WORKLOAD
+  namespace: NAMESPACE
+spec:
+  template:
+    spec:
+      containers:
+      - name: CONTAINER
+        resources:
+          limits:
+            device.microshift.io/serial: "1"
+```
+
+### API Extensions
+
+By including Generic Device Plugin (GDP) as code import, MicroShift configuration file
+will grow to accommodate the GDP's settings.
+
+Example config (based on default settings of GDP with changed `domain` and added `status`):
+```yaml
+genericDevicePlugin:
+  status: { Enabled , Disabled }
+  domain: device.microshift.io
+  devices:
```

Comment:
Any idea what are the limitations of this generic device plugin framework? What types of devices might not be well-suited?

Reply:
> this generic device plugin framework

"Generic Device Plugin" and "Device Plugin framework" are two different things.
"Generic DP" is a product, framework refers to https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/

As for "Generic DP": aAnything that needs some kind of special initialization.
For example it cannot do what, for example, [NVIDIA's device plugin](https://github.com/NVIDIA/k8s-device-plugin) does because it needs some unique procedures to prepare GPU before attaching (like resetting some queues (maybe that's networking gear) or clearing the memory so it doesn't read what previous pod was doing).

Reply:
Wait, you means "anything that *NOT* needs some kind of special initialization.

We might want to have a definition in the docs for this, some like:

Generice Device Plugin can handle all devices that are initialized and managed by the operating system and do not need any special initialization/procedures to a pod. 
Examples for generic devices that can be managed are serial ports, cameras, soundcards etc.
Example for special devices that required their own device plugin are e.g. GPUs, FPGAs etc.

Reply:
Yes, exactly

> What types of devices might not be well-suited?

> Anything that needs some kind of special initialization